third_party/libusb/webport/src/libusb_js_proxy.cc

```diff
@@ -1114,6 +1056,100 @@ int LibusbJsProxy::LibusbHandleEventsWithTimeout(libusb_context* context,
   return LIBUSB_SUCCESS;
 }
 
+std::function<void(GenericRequestResult)>
+LibusbJsProxy::MakeLibusbJsTransferCallback(
+    const std::string& js_api_name,
+    std::weak_ptr<libusb_context> context,
+    const UsbTransferDestination& transfer_destination,
+    TransferAsyncRequestStatePtr async_request_state) {
+  return [this, js_api_name, context, transfer_destination,
+          async_request_state](GenericRequestResult js_result) {
+    const std::shared_ptr<libusb_context> locked_context = context.lock();
+    if (!locked_context) {
+      // The context that was used for the original transfer submission has been
+      // destroyed already.
+      return;
+    }
+    if (!js_result.is_successful()) {
+      GOOGLE_SMART_CARD_LOG_INFO << js_api_name
+                                 << " failed: " << js_result.error_message();
+    }
+    LibusbJsTransferResult js_transfer_result;
+    RequestResult<LibusbJsTransferResult> converted_result =
+        RemoteCallAdaptor::ConvertResultPayload(
+            std::move(js_result), &js_transfer_result, &js_transfer_result);
+    if (transfer_destination.IsInputDirection()) {
+      locked_context->OnInputTransferResultReceived(
+          transfer_destination, std::move(converted_result));
+    } else {
+      locked_context->OnOutputTransferResultReceived(
+          async_request_state, std::move(converted_result));
+    }
+
+    // In case there are more transfers that were enqueued while waiting for an
+    // equivalent JS API call, a new call needs to be started.
+    MaybeStartTransferJsRequest(locked_context.get(), transfer_destination);
+  };
+}
+
+RemoteCallAsyncRequest LibusbJsProxy::PrepareTransferJsApiCall(
+    libusb_context* const context,
+    libusb_transfer* transfer,
+    const UsbTransferDestination& transfer_destination,
+    std::shared_ptr<TransferAsyncRequestState> async_request_state) {
+  const int64_t js_device_id =
+      transfer->dev_handle->device()->js_device().device_id;
+  const int64_t js_device_handle = transfer->dev_handle->js_device_handle();
+  const std::string js_api_name =
+      GetTransferJsApiName(static_cast<libusb_transfer_type>(transfer->type));
+
+  RemoteCallAsyncRequest async_request;
+  async_request.callback = MakeLibusbJsTransferCallback(
+      js_api_name, contexts_storage_.FindContextByAddress(context),
+      transfer_destination, async_request_state);
+
+  switch (transfer->type) {
+    case LIBUSB_TRANSFER_TYPE_CONTROL: {
+      LibusbJsControlTransferParameters transfer_params;
+      GOOGLE_SMART_CARD_CHECK(
```

Comment:
You're correct, this code will now crash when incorrect arguments are supplied instead of gracefully returning an error. I've decided to make the code stricter in this regard because we don't expect such incorrect arguments in our application and we'd rather learn about mistakes instead of letting them lost in some error handling mechanism.